src/session_impl.cpp

```diff
@@ -1807,16 +1805,68 @@ namespace aux {
 					// (which must be of the same family as the address we're
 					// connecting to)
 					if (device != ifs[k].name) continue;
+					eps.emplace_back(ifs[k].interface_address, port, device, ssl);
+				}
+			}
+		}
 
-					listen_socket_t const s = setup_listener(device
-						, tcp::endpoint(ifs[k].interface_address, port)
-						, flags | (ssl ? open_ssl_socket : 0), ec);
+		int const port_retries = m_settings.get_int(settings_pack::max_retry_port_bind);
 
-					if (!ec && s.sock)
-					{
-						m_listen_sockets.push_back(s);
-					}
-				}
+		// sockets we are keeping get moved to this list to prevent a socket from matching
+		// multiple endpoints
+		std::list<listen_socket_t> keep;
+
+		// remove any sockets which are no longer in the set of endpoints
+		// to listen on
+		// warning: O(n^2) operation!
+		// hopefully the system doesn't have too many interfaces
+		for (auto sock = m_listen_sockets.begin()
```

Comment:
do you think this would be simplified by using ``std::remove_if``? and the second pass could be to call close and actually erase the elements. I suppose that requires move support, or perhaps list nodes are swapped just with the chain pointers and stay stable